File: leetcode/142.linked-list-cycle-ii/cpp/detectcycle.cpp
struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution
{
public:
    ListNode *detectCycle(ListNode *head)
    {
        auto fast = head;
        auto slow = head;

        while (fast != nullptr && fast->next != nullptr)
        {
            fast = fast->next->next;
            slow = slow->next;

            if (fast == slow)
            {
                auto first = head;
                auto second = fast;

                while (first != second)
                {
                    first = first->next;
                    second = second->next;
                }

                return first;
            }
        }

        return nullptr;
    }
};

File: leetcode/142.linked-list-cycle-ii/go/detectcycle.go
package detectcycle

type ListNode struct {
	Val  int
	Next *ListNode
}

// N, 1
func detectCycle(head *ListNode) *ListNode {
	fast := head
	slow := head

	for fast != nil && fast.Next != nil {
		fast = fast.Next.Next
		slow = slow.Next

		if fast == slow {
			first := head
			second := fast

			for first != second {
				first = first.Next
				second = second.Next
			}

			return first
		}
	}

	return nil
}
